### ISSUE TITLE
Update /etc/hosts with $SET_HOSTNAME

### ISSUE THREAD

**Issue by baby-gnu** [OpenNebula/addon-context-linux#44]
Hello,

I just tried the 4.90.0 context for linux (ubuntu package) and found that `/etc/hosts` is not updated with the new host name provided by `SET_HOSTNAME` context attribute.

```
cat /etc/hosts
127.0.0.1   localhost
127.0.1.1   xenial.eole.lan xenial

# The following lines are desirable for IPv6 capable hosts
::1     localhost ip6-localhost ip6-loopback
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

I think the line 127.0.1.1 could be replaced.

Regards.

**Comment by FlorianHeigl** [comment]
127.0.1.1 is some Ubuntu speciality i think. 
@baby-gnu  What do you suggest to do on other OS? 
How about we assign it to the eth0 ip there?

**Comment by jfontan** [comment]
I think this is something that should be configured as in some cases makes a delay when logging in.

As Florian comments there are several ways to do this. I'm not sure what would be the best one. We can use `127.0.1.1` for Debian/Ubuntu and an interface IP for RHEL/CentOS. This should be configurable, just like [GATEWAY_IFACE](http://docs.opennebula.org/5.2/operation/references/template.html#context-section).

What do you think?

**Comment by baby-gnu** [comment]
I think that we can use the same solution for every distribution, no?

What does `ping 127.0.1.1` on RHEL/CentOS?

`127.0.0.0/8` is for loopback, so even if the IP address is not explicitly configured on lo, you should be able to ping it.

Regards.

**Comment by FlorianHeigl** [comment]
Hi, I checked;

```
       The recommended method of setting the FQDN is to make the hostname be an alias for the fully qualified name using /etc/hosts, DNS, or NIS.  For
       example, if the hostname was "ursula", one might have a line in /etc/hosts which reads

              127.0.1.1    ursula.example.com ursula
```

this is from CentOS7, and that part of the man page is very few lines away from talking about the decades-old host.conf mess. As far as I understand that man page it means we can expect to safely use that and the hostname command will know what it means.

**Comment by FlorianHeigl** [comment]
My updated scripts looked like this:
https://gist.github.com/FlorianHeigl/af4974cf988c97d25db7957fc2802d86 [debian]
https://gist.github.com/FlorianHeigl/52813ec9804ecb7b995525c1f3eabc80 [rhel]

**Comment by jfontan** [comment]
If I understand it correctly you add the name to the `eth0` IP and 127.0.1.1?

**Comment by FlorianHeigl** [comment]
@jfontan iirc for ubuntu i went with 127.0.1.1 and for redhat with eth0 ip.

An update:
Guys, I've been trying to get more info about the 127.0.1.1 story and it's impact.
One thing that someone told me just now.
He said mentioned he had various problems with this setup and I asked _which_. 

Actually it is pretty obvious:

If anyone binds their application to eth0, not to all interfaces, it'll become unreachable with the 127.0.1.1 host entry.

I'm not sure why I've not yet thougt about that, but I think it's a pretty valid concern.
So please give your input.

- 127.0.1.1 hack is easy to do and more reliable to configure
- using the eth0 IP seems safer

What could solve both things is the following algorithm:

round 1:
1. look at SET_HOSTNAME
2. if set, look if it has an identical DNS reverse entry on the eth0
3. if yes, go to A
3. if no, go to B
2. if not set, look at DHCP_HOSTNAME
3. if set, resolve reverse IP. 
4. if not resolved, go to C
4. if resolved, go to A


A. put an entry for eth0
B. put an entry for 127.0.1.1
C. do not touch /etc/hosts at all

**Comment by vholer** [comment]
Hi @FlorianHeigl, I have prepared a changes o

**Comment by FlorianHeigl** [comment]
It looks good - let's test it ;-)

@jfontan I noticed we have inconsistent variables now, (NAME vs name vs maybe some others).
What's the standard we should stick to?

**Comment by vholer** [comment]
Hi @FlorianHeigl, I have prepared few changes into `15-hostname` script based on your notes. Your proposed changes in last comment are reasonable. Not very tested, 

I think the contextualization shouldn't be very complicated, with any sophisticated logic behind. It should just prepare virtual machine for initial use and leave the rest to other tools (like Puppet or Ansible).

**Comment by vholer** [comment]
@FlorianHeigl Sorry for few incomplete comments, my keyboard has just its own head.
Regarding the upper- and lower-cased variable names, since the upper-cased variables are the context variables, I chose to use strictly lower-cased variable names for locals/other. Still playing with that, so your opinions are more than welcome. Thanks.

**Comment by FlorianHeigl** [comment]
@vholer I agree it should be easy - but
1) 100% correct if possible. Like the ansible_fqdn issue my own CM tool also completely fails to register the node if the name is inconclusive
2) in cloud envs there's quite a few use cases where people don't use any CM but do more via the "instance" metadata. I.e. they might really just re-deploy to update a setting.

We probably have to serve both cases. At least with working host resolution ;-)
Just to make sure you don't get me wrong: If the algorithm is sound god forbid we make it any more complicated. I just wanted to leave no loose ends.

**Comment by FlorianHeigl** [comment]
@vholer haha, yeah too many hotkeys in github nowadays :-)

your interpretion for UPPER/lower makes sense. I didn't notice, at least not now.

**Comment by FlorianHeigl** [comment]
@vholer I made a calendar note for testing on friday.
Which OS can you try?
You're on Ubuntu, right? Could you do 14 and 16? Can you also do Debian?
Then I'll try the CentOS / Oracle world (6+7)

SUSE I'm not sure I can do quick enough. I have already looked into it, will happen at some point.

**Comment by vholer** [comment]
@FlorianHeigl Sure, we can test it, that would be great. For the beginning, I think the main platforms supported by the ON (Debian, Ubuntu, and EL) are enough. What process do you suggest?

**Comment by vholer** [comment]
Hi @FlorianHeigl, not to spam people watching this repository with private communication, I'm available on Google Hangout/e-mail vlastimil.holer@gmail.com on Friday morning since about 9am. We can discuss the testing related things there. Thank you.

**Comment by vholer** [comment]
The functionality for setting the host and domain names has been merged via PR #64. I'm closing this issue right now. If the new behavior doesn't match all your requirements, please open a new issue.
